File: src/utils/scheduler.ts
import { Workplace, Employee, Availability, DailySchedule, Shift } from '../types';
import { differenceInDays, parseISO, format, addDays } from 'date-fns';

// Меньше = лучше. Три категории, см. правила выбора кандидата.
export function getCandidateScore(empPriority: number, skillPriority: number): number {
  const emp12 = empPriority === 1 || empPriority === 2;
  // Категория 1: emp 1/2, skill 1/2 — emp важнее skill
  if (emp12 && (skillPriority === 1 || skillPriority === 2)) {
    return (empPriority - 1) * 2 + (skillPriority - 1); // 0..3
  }
  // Категория 2: emp 1/2, skill 3 — по emp
  if (emp12) {
    return 10 + (empPriority - 1); // 10..11
  }
  // Категория 3: emp 3, любой skill — по skill
  return 20 + (skillPriority - 1); // 20..22
}

export function generateSchedule(
  workplaces: Workplace[],
  employees: Employee[],
  availabilities: Availability[],
  startDate: string,
  endDate: string
): DailySchedule[] {
  const start = parseISO(startDate);
  const end = parseISO(endDate);
  const daysCount = differenceInDays(end, start) + 1;
  const schedule: DailySchedule[] = [];

  const employeeShiftsCount: Record<string, number> = {};
  employees.forEach((e) => (employeeShiftsCount[e.id] = 0));

  for (let i = 0; i < daysCount; i++) {
    const currentDate = format(addDays(start, i), 'yyyy-MM-dd');
    const dayAssignments: Shift[] = [];
    const availableTodayIds = availabilities
      .filter((a) => a.date === currentDate)
      .map((a) => a.employeeId);
    const assignedToday = new Set<string>();

    // 1. Prepare all slots for the day
    interface Slot {
      workplaceId: string;
      roleId: string;
      slotIndex: number;
    }
    
    let openSlots: Slot[] = [];
    workplaces.forEach(wp => {
      wp.roles.forEach(role => {
        for (let c = 0; c < role.count; c++) {
          openSlots.push({ workplaceId: wp.id, roleId: role.id, slotIndex: c });
        }
      });
    });

    // 2. Greedy approach with "First-Fail" heuristics
    while (openSlots.length > 0) {
      // For each open slot, find potential candidates
      const slotCandidates = openSlots.map(slot => {
        const candidates = employees
          .filter(emp => {
            if (assignedToday.has(emp.id)) return false;
            if (!availableTodayIds.includes(emp.id)) return false;
            return emp.skills.some(s => s.workplaceId === slot.workplaceId && s.roleId === slot.roleId);
          })
          .map(emp => {
            const skill = emp.skills.find(s => s.workplaceId === slot.workplaceId && s.roleId === slot.roleId)!;
            const empPriority = emp.priority || 1;
            return {
              emp,
              skillPriority: skill.priority,
              empPriority,
              score: getCandidateScore(empPriority, skill.priority)
            };
          });
        return { slot, candidates };
      });

      const fillableSlots = slotCandidates.filter(sc => sc.candidates.length > 0);
      if (fillableSlots.length === 0) break;

      // Pick the MOST CONSTRAINED SLOT (the one with fewest candidates)
      fillableSlots.sort((a, b) => a.candidates.length - b.candidates.length);
      const target = fillableSlots[0];
      
      // Pick the BEST CANDIDATE for this slot: by score, then fewer shifts
      target.candidates.sort((a, b) => {
        if (a.score !== b.score) return a.score - b.score;
        return employeeShiftsCount[a.emp.id] - employeeShiftsCount[b.emp.id];
      });

      const selected = target.candidates[0];
      const note = availabilities.find(
        a => a.employeeId === selected.emp.id && a.date === currentDate
      )?.timeNote;

      // Assign
      dayAssignments.push({
        workplaceId: target.slot.workplaceId,
        roleId: target.slot.roleId,
        employeeId: selected.emp.id,
        priority: selected.skillPriority,
        ...(note ? { timeNote: note } : {})
      });

      assignedToday.add(selected.emp.id);
      employeeShiftsCount[selected.emp.id]++;
      
      // Remove this slot from openSlots
      openSlots = openSlots.filter(s => s !== target.slot);
    }

    schedule.push({
      date: currentDate,
      assignments: dayAssignments,
    });
  }

  return schedule;
}

File: bot/services/scheduler.py
"""Порт алгоритма расписания из src/utils/scheduler.ts под нормализованные данные.

data — db.get_full_data(): {"workplaces": [...roles...], "employees": [...skills...]}
availabilities — parse_availability(): [{"employee_name", "date", "time_note"}]
Матчинг сотрудника со слотом — по skill.role_id == role.id.
"""


def get_score(emp_priority: int, skill_priority: int) -> int:
    # Категория 1: empP 1-2, skillP 1-2 -> лучшие (0-11)
    if emp_priority <= 2 and skill_priority <= 2:
        return (emp_priority - 1) * 10 + (skill_priority - 1)
    # Категория 2: empP 1-2, skillP 3 -> средние (100-110)
    elif emp_priority <= 2:
        return 100 + (emp_priority - 1) * 10
    # Категория 3: empP 3 -> последние (200-220)
    else:
        return 200 + (skill_priority - 1) * 10


def generate_schedule(data: dict, availabilities: list[dict]) -> list[dict]:
    workplaces = data.get("workplaces", [])
    employees = data.get("employees", [])
    emp_by_name = {e["name"]: e for e in employees}

    dates = sorted({a["date"] for a in availabilities})
    # (name, date) -> time_note
    note_by = {(a["employee_name"], a["date"]): a.get("time_note") for a in availabilities}
    names_by_date = {}
    for a in availabilities:
        names_by_date.setdefault(a["date"], set()).add(a["employee_name"])

    shifts_count = {e["name"]: 0 for e in employees}
    schedule = []

    for day in dates:
        available = names_by_date.get(day, set()) & emp_by_name.keys()
        assigned = set()
        day_assignments = []

        open_slots = [
            {"wp": wp, "role": role}
            for wp in workplaces
            for role in wp.get("roles", [])
            for _ in range(role.get("count", 1))
        ]

        while open_slots:
            slot_candidates = []
            for slot in open_slots:
                role_id = slot["role"]["id"]
                cands = []
                for name in available:
                    if name in assigned:
                        continue
                    emp = emp_by_name[name]
                    skill = next((s for s in emp.get("skills", []) if s["role_id"] == role_id), None)
                    if not skill:
                        continue
                    emp_priority = emp.get("priority") or 1
                    cands.append({
                        "name": name,
                        "score": get_score(emp_priority, skill["priority"]),
                    })
                if cands:
                    slot_candidates.append({"slot": slot, "candidates": cands})

            if not slot_candidates:
                break

            # most-constrained slot first
            target = min(slot_candidates, key=lambda sc: len(sc["candidates"]))
            # лучший: меньший score, при равенстве — меньше смен за период
            best = min(target["candidates"], key=lambda c: (c["score"], shifts_count[c["name"]]))

            slot = target["slot"]
            day_assignments.append({
                "workplace_name": slot["wp"]["name"],
                "role_name": slot["role"]["name"],
                "employee_name": best["name"],
                "time_note": note_by.get((best["name"], day)),
                "score": best["score"],
            })

            assigned.add(best["name"])
            shifts_count[best["name"]] += 1
            open_slots.remove(slot)

        schedule.append({"date": day, "assignments": day_assignments})

    return schedule


if __name__ == "__main__":
    # категории строго упорядочены
    assert get_score(1, 1) < get_score(2, 2) < get_score(1, 3) < get_score(3, 1) < get_score(3, 3)

    data = {
        "workplaces": [{"id": "w1", "name": "Склад А", "roles": [
            {"id": "r1", "name": "Грузчик", "count": 1, "workplace_id": "w1"},
            {"id": "r2", "name": "Кладовщик", "count": 1, "workplace_id": "w1"},
        ]}],
        "employees": [
            {"id": "e1", "name": "Иван", "priority": 1, "skills": [
                {"role_id": "r1", "workplace_id": "w1", "priority": 1},
                {"role_id": "r2", "workplace_id": "w1", "priority": 1}]},
            {"id": "e2", "name": "Петр", "priority": 3, "skills": [
                {"role_id": "r1", "workplace_id": "w1", "priority": 1}]},
        ],
    }
    avail = [
        {"employee_name": "Иван", "date": "2025-07-01", "time_note": "с 15:00"},
        {"employee_name": "Петр", "date": "2025-07-01", "time_note": None},
    ]
    out = generate_schedule(data, avail)
    assert len(out) == 1 and out[0]["date"] == "2025-07-01"
    a = out[0]["assignments"]
    assert len(a) == 2  # оба слота закрыты

    # most-constrained: r2 (только Иван) -> Иван; r1 -> Петр. Один сотрудник = один слот.
    by_role = {x["role_name"]: x for x in a}
    assert by_role["Кладовщик"]["employee_name"] == "Иван"
    assert by_role["Кладовщик"]["time_note"] == "с 15:00"  # note скопирован
    assert by_role["Кладовщик"]["score"] == 0
    assert by_role["Грузчик"]["employee_name"] == "Петр"
    assert {x["employee_name"] for x in a} == {"Иван", "Петр"}  # без дублей

    print("ok")
